Add Chebyshev basis

nilsvu opened the pull request:
## Proposed changes

This PR adds Chebyshev polynomials as a choice for `Spectral::Basis`. This is mostly for testing that our `Spectral` interface works as expected also for a basis that is not `Spectral::Basis::Legendre`.

Indeed the interface needs a few additions to accommodate for basis functions with a non-unity weight function, i.e. the function w.r.t. which the basis functions are orthogonal. In particular, quadrature and integration is not identical in that case, so this PR makes explicit that `Spectral::quadrature_weights` returns the coefficients for computing the **definite integral** of a function (as opposed to the coefficients for computing the quadrature, i.e. the definite integral of a function times the weight function).

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add Chebyshev polynomial basis with Gauss and Gauss-Lobatto quadrature

nilsvu commented:
The commits are listed in the wrong order. It should be:

- 0cd3f6a
- eee8dea
- 91aa977
- cc451d7

geoffrey4444 commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+template <>
+double compute_basis_function_value<Basis::Chebyshev>(const size_t k,
+                                                      const double x) noexcept {
+  // Algorithm 21 from Kopriva's book, p. 60
```
Please include a full citation to Kopriva's book.

nilsvu commented:
Regarding what we just discussed at the meeting:

The `quadrature_weights` are only needed in `Spectral.cpp` and the `Spectral` tests. Elsewhere we only need `integral_weights`, and this also only for `definite_integral`. So should I remove `quadrature_weights` from `Spectral.hpp` and instead forward-declare where it's needed?

teukolsky commented:
In response to your question about integral_weights and quadrature_weights, this naming scheme is potentially confusing. In the DG literature, where the Legedre basis is used precisely because its weight function is unity, everyone talks about doing integrals by Gaussian quadrature and calls the weights "quadrature weights". I have never seen the term "integral weights" used. Most textbooks just tell you to check if someone is providing weights to integrate f(x) or f(x)w(x), where w(x) is the weight function. So I don't like adding the baggage of "integral_weights" anywhere. I would like to stick with only "quadrature_weights" and localize the definite integral changes in the Chebyshev class without polluting the Legendre class. Is this possible? Perhaps define definite integral to include the weight function for Chebyshev only?

nilsvu commented:
Thanks for following up on this @teukolsky, I'll make it work as you suggested :)

nilsvu commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -157,28 +241,91 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ExactQuadrature",
+                        Spectral::Quadrature::GaussLobatto>(
+        [](const size_t n) { return 2 * n - 3; });
+  }
+  // Is there a bound on the error of Chebyshev integrals that we can test?
```
Can anyone point me to literature on this? I'm only finding bounds on $\int f(x)w(x)dx$, which is what non-Legendre quadrature would be only used for of course and what I check in the test above. But would be nice to also test that $\int f(x)dx$ lies within expected limits too.

teukolsky commented:
Hi Nils,

I don't think Chebyshev definite integrals are ever actually used except 
with the implicit weight w(x). In that case the exactness for 
polynomials of degree N goes through. I don't know of any error bounds 
without the weight.

This makes me realize I don't understand how SpEC handles this - there, 
Chebyshev bases are used all the time.

Best,

Saul


On 5/27/2018 2:55 PM, Nils Leif Fischer wrote:
>
> *@nilsleiffischer* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp 
> <https://github.com/sxs-collaboration/spectre/pull/716#discussion_r191088980>:
>
> > +SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ExactIntegral",
> +                  "[NumericalAlgorithms][Spectral][Unit]") {
> +  // Legendre integrals are identical to quadrature since they have unit weight.
> +  SECTION(
> +      "Legendre-Gauss integral is exact to polynomial order 2*num_points-1") {
> +    test_exact_integral<Spectral::Basis::Legendre, Spectral::Quadrature::Gauss>(
> +        [](const size_t n) { return 2 * n - 1; });
> +  }
> +  SECTION(
> +      "Legendre-Gauss-Lobatto integral is exact to polynomial order "
> +      "2*num_points-3") {
> +    test_exact_integral<Spectral::Basis::Legendre,
> +                        Spectral::Quadrature::GaussLobatto>(
> +        [](const size_t n) { return 2 * n - 3; });
> +  }
> +  // Is there a bound on the error of Chebyshev integrals that we can test?
>
> Can anyone point me to literature on this? I'm only finding bounds on 
> $\int f(x)w(x)dx$...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sxs-collaboration/spectre/pull/716#pullrequestreview-123596487>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AeYvp_aFYo4S2-C7ElinkGII6-GGWEoJks5t2yDOgaJpZM4UHW8X>.
>



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

nilsvu commented:
Alright so I first pushed a commit testing that $\int f(x)/w(x)dx$ is integrated exactly for polynomials up to some order, complementing the test that the quadrature $\int f(x)w(x)dx$ is exact. But then I changed my mind and replaced the integral test with one that just checks that the `quadrature_weights` are the ones for integrals, i.e. $w_k/w(\xi_k)$. We could still test more integrals in `Test_DefiniteIntegral.cpp` later. Which tests do you think we should keep?

nilsvu commented:
Also I thought we may not want to have a `definite_integral` function at all for any basis other than Legendre, or rename it `quadrature` instead. Depends on whether we'll ever actually want to do integrals in a non-Legendre basis. For instance will we need surface integrals over a sphere, i.e. Lebedev quadrature? I haven't seen this in in YlmSpherepack.hpp just now, but might have missed it.

teukolsky commented:
I don't know much about Lebedev quadrature for spheres, but I'm not convinced it's any better than the methods in Spherepack, which essentially do two 1-d integrations. The power of Gaussian quadrature is *not* that it integrates polynomials exactly up to a higher order than one might have expected, which is what the textbooks all emphasize. It is that the convergence of the quadrature is exponential for smooth functions. As far as I can tell, Lebedev isn't a Gaussian quadrature in that sense (I may be wrong). So if we ever have spherical domains, we might as well stick with Spherepack.

nilsvu commented:
Alright so I guess what we need to decide for this PR is whether we want to have a `definite_integral` function that computes $\int f(x) dx$ no matter the basis, or a `quadrature` that computes $\int f(x) w(x) dx$, or both. This is of course irrelevant for Legendre meshes (except for the function name).

Since I guess we'd only ever want to use a non-Legendre basis for instances where weighted integrals occur, e.g. spherical surface integrals $\int f(x) d\Omega$ or Chebyshev $\int f(x) dx/sqrt(1-x^2)$, it could be the `quadrature` that we (always) want. Does this make sense? I'd also be interested in how this is handled in SpEC (discussed this also with @haraldp271 today).

Could we make a decision on this, so that we can finish up this PR?

kidder commented:
I have been busy writing a proposal the past few days, but will look at what SpEC does and give you my opinion sometime this weekend

nilsvu commented:
Sounds good 👍

kidder commented:
In the spectral namespace just want the definite_integral function that integrates the function in logical coordinates.  For SpEC this is done by a product of nodal values and something we called quadrature weights

nilsvu commented:
Done as you suggested @kidder. Should be ready.

kidder commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -17,6 +18,88 @@
 
 namespace {
 
+auto evaluate_polynomial = [](const std::vector<double>& coeffs,
```
just use functions instead of lambdas and give the explicit return type

kidder commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -51,18 +52,23 @@ auto polynomial_integral = [](const size_t deg) {
 };
 
 template <Spectral::Basis BasisType, Spectral::Quadrature QuadratureType>
```
add a template parameter F to use instead of `std::function...` below

nilsvu replied:
why is that better? appears less precise to me. This function expects a `size_t` for its argument.

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+  // Since the trigonometric form is expensive to compute we make use of it only
+  // for large k. It remains to be determined what is a reasonable crossover
+  // point between the two forms, e.g. depending on compiler settings.
+  static constexpr size_t chebyshev_recursion_to_trigonometry_crossover = 10;
```
Is this comment yours or from the Kopriva reference?  Why is the trigonometric form used for large k?

nilsvu replied:
Kopriva comments on this in the book (not in these exact words). Up to some `k` the recursion is faster than computing the `cos` and `acos`. At which `k` it becomes more efficient to compute the trigonometric form, as opposed to performing the recursion, varies a lot, so Kopriva recommends testing this for every compiler and computer architecture. Since this is not of high priority right now I added the comment that this is to be determined.
Added a reference to the section in the book where this is discussed now, too.

wthrowe replied:
The result of this function is cached.  Marginal speed improvements are not worth complicating the logic here.  Just pick one method.

teukolsky replied:
I agree. For the (small) values of k we will be using, recursion is the way to go.

nilsvu replied:
sounds good 👍

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+  DataVector x(num_points);
+  const DataVector w(num_points, M_PI / num_points);
+  for (size_t j = 0; j < num_points; j++) {
+    x[j] = -cos((2. * double(j) + 1.) / (2. * poly_degree + 2.) * M_PI);
```
don't need the cast to double...     
`-cos(M_PI_2 *(2.*j+1.)/(poly_degree+1))`

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+  DataVector x(num_points);
+  DataVector w(num_points, M_PI / poly_degree);
+  for (size_t j = 0; j < num_points; j++) {
+    x[j] = -cos(double(j) / double(poly_degree) * M_PI);
```
`M_PI * j / poly_degree`

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+  for (size_t j = 0; j < num_points; j++) {
+    x[j] = -cos(double(j) / double(poly_degree) * M_PI);
+  }
+  w[0] /= 2.;
```
`*= 0.5`

nilsvu replied:
ok but why :D

kidder replied:
multiplication is much faster than division

nilsvu replied:
is this not something the compiler can cope with? anyway, changed it

wthrowe replied:
It can in theory handle it as long as the LHS is a fundamental and the RHS is a literal power of two.  (I expect compilers actually do do that transformation but I haven't checked.)  These conditions are met in this case, but it is easier to remember and check a rule of "never divide by a literal double" than the rule with the two part exception above, and as computational physicists we are used to interpreting 0.5, 0.25, and similar simple fractions.

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+    x[j] = -cos(double(j) / double(poly_degree) * M_PI);
+  }
+  w[0] /= 2.;
+  w[num_points - 1] /= 2.;
```
again

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,121 @@
+         "Chebyshev-Gauss quadrature requires at least one collocation point.");
+  const size_t poly_degree = num_points - 1;
+  DataVector x(num_points);
+  const DataVector w(num_points, M_PI / num_points);
```
remove const and move below

kidder commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGauss.cpp:
```diff
@@ -0,0 +1,81 @@
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+namespace Spectral {
+// To test the \f$w_k\f$ as opposed to the Spectral::quadrature_weights. We
```
remove `\f$` as this is not a doxygen comment

nilsvu replied:
I'd prefer to mark up comments the same as doxygen comments for consistency, this also makes it easier to move comments around between doxygen-formatted environments and others. If you're strongly opposed I'll change it though.

nilsdeppe replied:
In that case you should put `Spectral::quadrature` weights in ticks :)

nilsvu replied:
Not sure what’s best practice here. I follow the Doxygen documentation where they don’t put types and functions in backticks because they are parsed and linked: https://www.stack.nl/~dimitri/doxygen/manual/autolink.html

nilsdeppe replied:
We use backticks literally everywhere in the Doxygen documentation...

kidder commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGauss.cpp:
```diff
@@ -0,0 +1,81 @@
+
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ChebyshevGauss.PointsAndWeights",
+                  "[NumericalAlgorithms][Spectral][Unit]") {
+  // Compare Chebyshev-Gauss points to Mathematica code available on request
```
could your code to do this be written in python instead of Mathematica?

nilsvu replied:
Yes. In this case these are simple analytic expressions though, so I just added them to the comment with a reference.

kidder commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGauss.cpp:
```diff
@@ -0,0 +1,81 @@
+      Spectral::collocation_points<Spectral::Basis::Chebyshev,
+                                   Spectral::Quadrature::Gauss>(num_points);
+  CHECK_ITERABLE_APPROX(expected_points, points);
+  // We test the \f$w_k\f$ here directly. Test_Spectral.cpp and
```
Is it necessary to test a private function?  Presumably something else would fail if it was implemented incorrectly...

nilsvu replied:
Yes `definite_integral` should fail, and a lot of other tests too. This tests that the algorithm to compute points and weights works as expected, which would be hard to debug otherwise. So I think this test is useful. This function is only private because we decided not to expose the w_k but only the w_k/w(x_k) for computing definite integrals.

kidder commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -119,3 +119,10 @@ std::pair<DataVector, DataVector> compute_collocation_points_and_weights<
 
 }  // namespace Spectral
+
+/*
```
This won't show up in the documentation.  Perhaps put a citation in the hpp file, and don't use \cite in a non-doxygen comment

nilsvu replied:
Hm these Kopriva citations only appear in cpp files since they provide algorithm implementations. They don't even necessarily have an hpp. I thought adding this _References_ section to the bottom of the file where the citation appears and using the doxygen \cite for both making clear that this is a citation and allowing for the comment to be moved somewhere else eventually made sense. If you have another suggestion on how to do things I can change it.

nilsdeppe replied:
You've probably figured out how I feel about this :) Hopefully what I've provided as alternatives is acceptable for you.

kidder reviewed (commented):
you can squash

nilsvu committed:
Consolidate Spectral tests to prepare for adding Chebyshev tests

nilsvu committed:
Add test for exact differentiation

nilsvu committed:
Accomodate for basis functions with non-unity weight

We choose to let Spectral::quadrature_weights return the weights for computing
definite integrals, as opposed to those for computing the weighted inner product.

nilsvu commented:
done

kidder commented:
@nilsdeppe please look at

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -19,3 +22,16 @@ SPECTRE_ALWAYS_INLINE T number_of_digits(const T number) {
+ * \brief Evaluate a polynomial with Horner's rule
+ */
+template <typename T>
+T evaluate_polynomial(const std::vector<double>& coeffs, const T& x) {
```
`noexcept` and needs tests for `double` and presumably you also had `DataVector` in mind when writing this as a template.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -8,34 +8,57 @@
+namespace {
+
+DataVector unit_polynomial(const size_t deg, const DataVector& x) {
+  // Simply choose all polynomial coefficients to one
```
`to be one` and remove `Simply`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -96,60 +107,38 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ExactInterpolation",
-      func_value += pow(x, p) / double(p);
+template <Spectral::Basis BasisType, Spectral::Quadrature QuadratureType,
+          typename Function>
+void test_exact_quadrature(const Function& max_poly_deg) {
```
Is this not testing integration? I find the name `exact_quadrature` to not tell me anything at all

nilsvu replied:
It tests that the quadrature, i.e. the weighted inner product, is exact to the expected polynomial order. There‘s no such test for integration for non-unit weight basis functions.

nilsdeppe replied:
I'm confused now. Didn't you change the weights so that `sum_k w_k f_k` is an approximation to the integral?

nilsvu replied:
No the terminology is explained in the documentation of `Spectral::quadrature_weights`, it returns w_k/w(x_k). I think it’s confusing too, which is why I called those `Spectral::integral_weights` before. However, @teukolsky pointed out that this is not what they are referred to in the literature (see discussion above), so we chose this naming scheme.

nilsdeppe replied:
Thanks for explaining and I'm glad there's documentation explaining this.

nilsdeppe commented on src/Utilities/Math.hpp:
```diff
@@ -19,3 +22,16 @@ SPECTRE_ALWAYS_INLINE T number_of_digits(const T number) {
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Evaluate a polynomial with Horner's rule
```
You should list the requirements on the types here

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Legendre.cpp:
```diff
@@ -21,20 +21,20 @@ namespace Spectral {
 
```
Your commit message doesn't tell me *why*

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Spectral.hpp:
```diff
@@ -93,7 +93,23 @@ const DataVector& collocation_points(size_t num_points) noexcept;
+ * f(x)\mathrm{d}x\f$.
+ *
+ * Note that the term _quadrature_ also often refers to the quantity
+ * \f$Q[f]=\int f(x)w(x)\mathrm{d}x\approx f_k w_k\f$. Here, \f$w(x)\f$
```
Should there not be a sum over `k` here?

nilsvu replied:
Yeah it‘s implied, I‘ll add it if you think that‘s unclear.

nilsdeppe replied:
It is unclear because it's never written this way in text books or anywhere. I've yet to see anybody write it this way

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -0,0 +1,118 @@
+    default:
+      // These values can be computed either through recursion
+      // (implemented here), or analytically as `cos(k * acos(x))`.
+      // Since the trigonometric form is expensive to compute it is useful only
```
You should figure out what "large" means and add an assert. I think anything that is less accurate than 1e-13 or 14 relative error is bad. Also, isn't this something we cache? If so, just compute it the accurate way :)

teukolsky replied:
The trig method will never be useful for us - it's only faster than recurrence when N is many dozens, and the recurrence is accurate for N in the many thousands. I think it's fine for us not to bother with an assert here.

nilsdeppe replied:
I'm now confused. The C++ comment makes it sound like we don't do trig because it's expensive but is also more accurate. I'm interpreting your response as saying the recursion is more accurate, which confuses me.

I still think if we cache these values we should just use the trig form since any "cost" is basically zero anyway.

teukolsky replied:
I'm not sure where the claim that the trig form is more accurate comes from. Trefethen has a whole book and software package for computing with Chebyshev polynomials and uses recurrence to compute them. The only place I've seen the trig form used is in Boyd's book, which is over 30 years old. Bottom line: use recurrences, they're faster and more accurate. Anyway, we're "never" going to use Chebyshev polynomials for DG.

nilsdeppe replied:
Ah okay, then the comment needs to be update so that it is actually correct. Thanks for the clarification! :)

nilsvu replied:
The comment is actually correct. There‘s also a reference to more information. Is it the wording `analytic expression` that gives you the impression that it would be more accurate?

nilsdeppe replied:
I don't see how it wouldn't be more accurate for this case, or at least just as accurate. The comment is very misleading and implying there are accuracy reasons for switching from recursion. It definitely needs to be corrected.

nilsvu replied:
I don‘t understand, the comment only says that the trigonometric form is more expensive.

nilsdeppe replied:
No, it gives a reason not to use it under certain conditions, not explaining why those conditions are imposed and when, leaving it to the reader to assume what is meant.

teukolsky replied:
I looked at the reference (Kopriva). He says the crossover point in terms of speed is around N=70. So let's stick wtih recursion and leave the comment as is.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGauss.cpp:
```diff
@@ -0,0 +1,90 @@
+  CHECK_ITERABLE_APPROX(expected_points, points);
+  // We test the \f$w_k\f$ here directly. Test_Spectral.cpp and
+  // Test_DefiniteIntegral.cpp take care of testing the
+  // Spectral::quadrature_weights.
```
here too

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGauss.cpp:
```diff
@@ -0,0 +1,90 @@
+  // Compare Chebyshev-Gauss points and weights to the analytic expressions
+  // \f$\x_j=-\cos{\frac{2j+1}{2p+2}\pi}\f$ and \f$w_k=\frac{\pi}{N}\f$, where
+  // \f$p=N-1\f$ is the polynomial degree, \f$N\f$ the number of collocation
+  // points and \f$0\leq j \leq p\f$ (\cite Kopriva, eqn. (1.128)).
```
how does this \cite work since this is C++ and not latex? It seems rather silly what's been done here. Just put a reference on the `spectral` namespace that is complete and explains where the algorithms come from then call it `Kopriva`. If you really want there to be more inline then say `kopriva (10.1007/978-90-481-2261-5)`. The DOI and author name will be enough to uniquely identify the book.

nilsvu replied:
It would be best to cite a proper bibtex entry, so putting the reference in the namespace documentation sound like a good idea! Then it‘s actually parses by Doxygen so it shows up as a citation: https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcite

I‘d like to refer to the particular algorithms in the cpp files too, so how about there I leave the `\cite`s and update them with the correct bibtex entry?

nilsdeppe replied:
I think adding it to the namespace is good enough. In the cpp file just refer to it as `Kopriva algorithm blah`. If you really want something more direct in the cpp file do the DOI think. Nothing more is needed.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGaussLobatto.cpp:
```diff
@@ -0,0 +1,88 @@
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+namespace Spectral {
+// To test the \f$w_k\f$ as opposed to the Spectral::quadrature_weights. We
```
again

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGaussLobatto.cpp:
```diff
@@ -0,0 +1,88 @@
+  CHECK_ITERABLE_APPROX(expected_points, points);
+  // We test the \f$w_k\f$ here directly. Test_Spectral.cpp and
+  // Test_DefiniteIntegral.cpp take care of testing the
+  // Spectral::quadrature_weights.
```
here too

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_ChebyshevGaussLobatto.cpp:
```diff
@@ -0,0 +1,88 @@
+  // expressions \f$\x_j=-\cos{\frac{j\pi}{p}}\f$ and \f$w_k=\frac{\pi}{p}\f$
+  // except for \f$w_{0,p}=\frac{\pi}{2p}\f$, where \f$p=N-1\f$ is the
+  // polynomial degree, \f$N\f$ the number of collocation points and \f$0\leq j
+  // \leq p\f$ (\cite Kopriva, eqn. (1.130)).
```
same thing about `\cite`

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -224,6 +260,46 @@ SPECTRE_TEST_CASE("Unit.Numerical.Spectral.ExactQuadrature",
+  // For a function \f$f(x)\f$ the exact quadrature is
+  // \f$\int_{-1}^{1}f(x)w(x)\mathrm{d}x\f$ where \f$w(x)=1/sqrt(1-x^2)\f$ is
+  // the Chebyshev weight. We test this here for polynomials with unit
+  // coefficients \f$f(x)=1+x+x^2+...\f$.
```
Please use `\ldots`, otherwise switch to markdown ticks for these things.

nilsdeppe commented on src/NumericalAlgorithms/Spectral/Chebyshev.cpp:
```diff
@@ -23,7 +23,7 @@ DataVector compute_basis_function_values(size_t, const DataVector&) noexcept;
 template <>
 DataVector compute_basis_function_values<Basis::Chebyshev>(
     const size_t k, const DataVector& x) noexcept {
```
This is silly. Add an explanation to the namespace for where the algorithms come from then cite it just as` Kopriva`. If you really want more then do the `Kopriva (DOI)` thing. It's C++, not latex. Just drop this commit.

nilsdeppe commented:
@nilsleiffischer are you abandoning this?

nilsvu commented:
No it's just very low priority. In particular I want to improve how we do references by citing properly from a bibtex file. I might split that off into another PR though.
**Edit:** #878

nilsvu commented:
Adressed comments, had to rebase to reword the one commit message, but diffs don't change until the fixup commits.

Removed the `\cite`s and will add a proper reference to a bibtex file in the namespace documentation in another PR (#878). So this one should be ready to go.

nilsvu committed:
Make non-cached spectral functions public, closes #801

kidder commented:
@nilsleiffischer this isn't compiling...

nilsvu commented:
Fixed it. Also squashed to allow #843 to cherry-pick `evaluate_polynomial`.

nilsvu commented:
rebased on develop for the `evaluate_polynomial` implementation

kidder commented:
travis failures are timeouts
